### PR TITLE
Fix #1592: Implement procedure sp_getapplock for locking activations on MSSQL

### LINKED ISSUE #1592: Implement procedure sp_getapplock for locking activations on MSSQL
In order to prevent deadlocks due to pesimmistic locking of activations and operations on MSSQL, we should implement the `sp_getapplock` to lock these database resources in a more reliable way on MSSQL.

### PR DESCRIPTION
**Pull request by romanstrobl**

### CODE REVIEW

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/conditions/IsNotMssqlCondition.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+public class IsNotMssqlCondition implements Condition {
+
+    @Override
+    public boolean matches(ConditionContext context, @NonNull AnnotatedTypeMetadata metadata) {
+        final Environment environment = Objects.requireNonNull(context.getBeanFactory()).getBean(Environment.class);
+        final String url = environment.getProperty("spring.datasource.url");
+        return DatabaseDriver.fromJdbcUrl(url) != DatabaseDriver.SQLSERVER;
+    }
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/conditions/IsNotMssqlCondition.java:43]: 🤔 We duplicate the whole logic with only a different comparison. There should be a way how to negate the condition, something like this.
```suggestion
public class IsNotMssqlCondition extends NoneOfNestedConditions {

    @IsMssqlCondition()
    static class IsMssql {
    }
```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/repository/OperationRepository.java [hunk 2]
```diff
@@ -45,6 +45,37 @@ public interface OperationRepository extends CrudRepository<OperationEntity, Str
...
+     * The operation is locked using stored procedure sp_getapplock in exclusive mode.
+     * The lock is released automatically at the end of the transaction. Transaction isolation
+     * level READ COMMITTED is used because the lock is pessimistic, optimistic locking would
+     * cause an UPDATE conflict error. The stored procedure raises an error in case the lock
+     * could not be acquired.
+     *
+     * @param operationId Operation ID
+     * @return Operation with given ID
+     */
+    @Query(value = "BEGIN TRANSACTION;\n" +
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/repository/OperationRepository.java:59]: May we use the text block?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ActivationServiceBehavior.java [hunk 3]
```diff
@@ -2479,7 +2480,11 @@ public void expireActivations() {
         try (final Stream<ActivationRecordEntity> abandonedActivations = activationRepository.findAbandonedActivations(activationStatuses, lookBackTimestamp, currentTimestamp)) {
             abandonedActivations.forEach(activation -> {
                 logger.info("Removing abandoned activation with ID: {}", activation.getActivationId());
-                deactivatePendingActivation(currentTimestamp, activation, true);
+                try {
+                    deactivatePendingActivation(currentTimestamp, activation, false);
+                } catch (GenericServiceException e) {
+                    logger.error("Activation not found, activation ID: {}", activation.getActivationId());
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ActivationServiceBehavior.java:2486]: 🤔 Is the log message correct? Should be something like `Activation was not possible to deactivate, activation ID: {}`?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthAssertionProvider.java [hunk 4]
```diff
@@ -173,9 +176,12 @@ public AssertionChallenge failAssertion(String challengeValue, AuthenticatorDeta
             operationFailApprovalRequest.setOperationId(operationId);
             operationFailApprovalRequest.getAdditionalData().putAll(prepareAdditionalData(authenticatorDetail, authenticatorData, clientDataJSON));
 
-            final ActivationRecordEntity activationWithLock = repositoryCatalogue.getActivationRepository().findActivationWithLock(authenticatorDetail.getActivationId());
+            final ActivationRecordEntity activation = activationQueryService.findActivationForUpdate(authenticatorDetail.getActivationId()).orElseThrow(() -> {
+                logger.info("Activation not found, activation ID: {}", authenticatorDetail.getActivationId());
+                return new Fido2AuthenticationFailedException("Activation not found");
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthAssertionProvider.java:181]: I suggest to keep the context in the exception.
```suggestion
                return new Fido2AuthenticationFailedException("Activation ID: %s not found".formatted(authenticatorDetail.getActivationId()));
```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryServiceMssql.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * Service for activation queries with pessimistic locking, MSSQL implementation.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Service
+@Conditional(IsMssqlCondition.class)
+public class ActivationQueryServiceMssql implements ActivationQueryService {
+
+    private static final Logger logger = LoggerFactory.getLogger(ActivationQueryServiceMssql.class);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryServiceMssql.java:44]: We may use `@Slf4j` instead.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryServiceDefault.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.util.Optional;
+
+/**
+ * Service for operation queries with pessimistic locking.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Service
+@Conditional(IsNotMssqlCondition.class)
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryServiceDefault.java:37]: 🤔 Have you considered doing this conditional configuration even on the repository level and keeping a single service?

**Review by banterCZ — APPROVED** [review]
Looks good.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/conditions/IsNotMssqlCondition.java (hunk 1 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/conditions/IsNotMssqlCondition.java:43]: Sure, we can do this.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryServiceDefault.java (hunk 6 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryServiceDefault.java:37]: I have considered it. If we do this for operations, we should do it for activations, too. However there will be many duplicate methods on `ActivationRepository`.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java [hunk 7]
```diff
@@ -120,15 +124,14 @@ public RegistrationChallenge findRegistrationChallengeByValue(String application
         final String activationCode = split[0];
 
         // Only allow created activations to be finished
-        final ActivationRecordEntity activationRecordEntity = repositoryCatalogue.getActivationRepository()
-                .findCreatedActivationWithoutLock(applicationId, activationCode, List.of(ActivationStatus.CREATED), currentTimestamp);
+        ActivationRecordEntity activation = activationQueryService.findActivationByCodeWithoutLock(applicationId, activationCode, List.of(ActivationStatus.CREATED), currentTimestamp).orElseThrow(() -> {
+            logger.warn("Activation with activation code: {} could not be obtained. It either does not exist or it already expired.", activationCode);
+            // Rollback is not required, error occurs before writing to database
+            return new Fido2AuthenticationFailedException("Activation could not be found");
```
Comment (zcgandcomp) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java:130]: the change of message is intentional?
"Activation failed" to "Activation could not be found"

**Review by zcgandcomp — APPROVED** [review]
The logic seems OK. The changes should not have an impact on the expected functionality.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java (hunk 7 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java:130]: I was trying to provide a more specific message, but we can revert to the original message, no problem.

**Review by banterCZ — APPROVED** [review]

**Review by banterCZ — DISMISSED** [review]

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java (hunk 7 above)
Comment (zcgandcomp) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java]: NP, just checking if this was intentional.

File: docs/Deploying-PowerAuth-Server.md [hunk 8]
```diff
@@ -234,6 +234,25 @@ spring.jpa.hibernate.ddl-auto=none
 spring.datasource.jndi-name=java:/jdbc/powerauth
 ```
 
+### Deploying on MSSQL Database
+
+The MSSQL database must run in SNAPSHOT transaction isolation mode for PowerAuth server.
```
Comment (banterCZ) [inline docs/Deploying-PowerAuth-Server.md:239]: I believe that we may use `<!-- begin box warning -->`

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java [hunk 9]
```diff
@@ -781,10 +787,15 @@ public OperationListResponse findAllOperationsForUser(final OperationListForUser
...
+            try (final Stream<OperationEntity> operationsForUser = operationQueryService.findAllOperationsForUser(userId, applicationIds, activationId, activationFlags.isEmpty() ? null : activationFlags, operationListRequest.pageable())) {
                 operationsForUser.forEach(op -> {
-                    final OperationEntity operationEntity = expireOperation(op, currentTimestamp);
-                    result.add(convertFromEntity(operationEntity));
+                    final OperationEntity operationEntity;
+                    try {
+                        operationEntity = expireOperation(op, currentTimestamp);
+                        result.add(convertFromEntity(operationEntity));
+                    } catch (GenericServiceException e) {
+                        logger.error("Operation expiration failed, operation ID: {}", op.getId());
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java:797]: A swallowed exception, let's log it at least for the DEBUG level.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java [hunk 10]
```diff
@@ -825,13 +836,18 @@ public OperationListResponse findPendingOperationsForUser(OperationListForUserRe
...
+                    final OperationEntity operationEntity;
+                    try {
+                        operationEntity = expireOperation(op, currentTimestamp);
+                        // Skip operation that just expired
+                        if (OperationStatusDo.PENDING.equals(operationEntity.getStatus())) {
+                            final OperationDetailResponse operationDetail = convertFromEntityAndFillOtp(operationEntity);
+                            result.add(operationDetail);
+                        }
+                    } catch (GenericServiceException e) {
+                        logger.error("Operation expiration failed, operation ID: {}", op.getId());
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java:850]: dtto

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java [hunk 11]
```diff
@@ -1167,8 +1192,14 @@ public void expireOperations() {
...
 
         final PageRequest pageRequest = PageRequest.of(0, powerAuthServiceConfiguration.getExpireOperationsLimit());
-        try (final Stream<OperationEntity> pendingOperations = operationRepository.findExpiredPendingOperations(currentTimestamp, pageRequest)) {
-            pendingOperations.forEach(op -> expireOperation(op, currentTimestamp));
+        try (final Stream<OperationEntity> pendingOperations = operationQueryService.findExpiredPendingOperations(currentTimestamp, pageRequest)) {
+            pendingOperations.forEach(op -> {
+                try {
+                    expireOperation(op, currentTimestamp);
+                } catch (GenericServiceException e) {
+                    logger.error("Operation expiration failed, operation ID: {}", op.getId());
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java:1200]: I do not like this pattern when the whole story is not described in a single log entry, but we may fine-tune it later on.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/VaultUnlockServiceBehavior.java [hunk 12]
```diff
@@ -162,16 +165,18 @@ public VaultUnlockResponse unlockVault(VaultUnlockRequest request) throws Generi
                 throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
             }
             // Lookup the activation
-            final ActivationRecordEntity activation = repositoryCatalogue.getActivationRepository().findActivationWithoutLock(activationId);
+            final Optional<ActivationRecordEntity> activationOptional = activationQueryService.findActivationWithoutLock(activationId);
 
             // Check if the activation is in correct state
-            if (activation == null || !ActivationStatus.ACTIVE.equals(activation.getActivationStatus())) {
+            if (activationOptional.isEmpty() || !ActivationStatus.ACTIVE.equals(activationOptional.get().getActivationStatus())) {
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/VaultUnlockServiceBehavior.java:171]: If already touched, we may get rid of `equals` and use `!=` for enum.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryService.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Find all activations for given user ID
+     *
+     * @param applicationId Application ID.
+     * @param externalId External identifier.
+     * @return List of activations for given user
+     */
+    List<ActivationRecordEntity> findByExternalId(String applicationId, String externalId);
+
+
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryService.java:110]: Two empty lines.

```suggestion
    List<ActivationRecordEntity> findByExternalId(String applicationId, String externalId);

```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryServiceDefault.java [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+
+    @Override
+    public Stream<ActivationRecordEntity> findAbandonedActivations(Collection<ActivationStatus> states, Date startingTimestamp, Date currentTimestamp) {
+        try {
+            return activationRepository.findAbandonedActivations(states, startingTimestamp, currentTimestamp);
+        } catch (Exception ex) {
+            logger.error("Activation query failed", ex);
+            return Stream.empty();
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryServiceDefault.java:118]: 🤔 We may replace it by a template method pattern or closure.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryService.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Service for operation queries with pessimistic locking.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+public interface OperationQueryService {
+
+    Optional<OperationEntity> findOperationForUpdate(String operationId);
+
+    Optional<OperationEntity> findOperationWithoutLock(String operationId);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryService.java:37]: May we have a JavaDoc at least for the first two methods?

**Review by banterCZ — DISMISSED** [review]
Approving. My comments may be addressed in a separate PR.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/VaultUnlockServiceBehavior.java (hunk 12 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/VaultUnlockServiceBehavior.java:171]: I will fix it in all occurrences.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryServiceDefault.java (hunk 14 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/ActivationQueryServiceDefault.java:118]: We can refactor it later.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryService.java (hunk 15 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/persistence/OperationQueryService.java:37]: Yes, I will write Javadoc for all methods.

**Review by banterCZ — APPROVED** [review]

**Comment by romanstrobl** [comment]
I am merging the pull request. We can resolved any remaining issues in separate issues.

**Pull request merged**